packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-asset-bucket.ts

```diff
@@ -0,0 +1,83 @@
+import * as cdk from '@aws-cdk/core';
+import { IBucket, BlockPublicAccess, Bucket } from '../../../aws-s3';
+import { BucketDeployment, ISource, Source } from '../../../aws-s3-deployment';
+import { Construct } from 'constructs';
+import { hashValues } from './util';
+
+/**
+ * A Service Catalog product stack asset bucket, which is similar in form to an Amazon S3 bucket.
+ * You can store multiple product stack assets and collectively deploy them to S3.
+ */
+export class ProductStackAssetBucket extends Construct {
+  private readonly bucketName: string;
+  private readonly bucket: IBucket;
+  private readonly assets: ISource[];
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.bucketName = this.generateBucketName(id);
+
+    this.bucket = new Bucket(scope, `${id}S3Bucket`, {
+      bucketName: this.bucketName,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+    });
+
+    this.assets = [];
+  }
+
+  /**
+   * Fetch the S3 bucket.
+   *
+   * @internal
+   */
+  public getBucket(): IBucket {
+    return this.bucket;
+  }
+
+  /**
+   * Generate unique name for S3 bucket.
+   *
+   * @internal
+   */
+  private generateBucketName(id: string): string {
+    const accountId = cdk.Stack.of(this).account;
+    if (cdk.Token.isUnresolved(accountId)) {
+      throw new Error('CDK Account ID must be defined in the application environment');
+    }
+    return `product-stack-asset-bucket-${accountId}-${hashValues(id)}`;
+  }
+
+  /**
+   * Fetch the expected S3 location of an asset.
+   *
+   * @internal
+   */
+  public addAsset(asset: cdk.FileAssetSource): cdk.FileAssetLocation {
+    const assetPath = './cdk.out/' + asset.fileName;
```

Comment:
This is making too many assumptions.

Instead, I would hand the asset to the underlying stack's `StackSynthesizer`, and hand the result of that to the `BucketDeployment` (as a `Source.s3()`).